File: first_year/sem2/Graphs/lab/assignment1/PY/directed_graph.py
from edge import Edge
from copy import deepcopy

class DirectedGraph:
    def __init__(self, n=0, m=0):
        self.__n = n
        self.__m = m
        self.__vertices = list()
        self.__outbounds = dict()
        self.__inbounds = dict()
        self.__costs = dict()

        for i in range(n):
            self.__vertices.append(i)
            self.__outbounds[i] = []
            self.__inbounds[i] = []

    def add_vertex(self, v:int):
        if self.valid_vertex(v) == False:
            raise Exception("Invalid vertex!\n")

        if self.existing_vertex(v) == True:
            raise Exception("Vertex already exists!\n")

        self.__vertices.append(v)
        self.__outbounds[v] = []
        self.__inbounds[v] = []
        self.__n = max(self.__n, v)
    
    def remove_vertex(self, v:int):
        if self.valid_vertex(v) == False:
            raise Exception("Invalid vertex!\n")
        
        if self.existing_vertex(v) == False:
            raise Exception("Vertex not found!\n")

        cnt = 0
        for neighbors in self.__outbounds.values():
            if v in neighbors:
                self.remove_edge(cnt, v)
            cnt += 1

        cnt = 0
        for neighbors in self.__inbounds.values():
            if v in neighbors:
                self.remove_edge(v, cnt)
            cnt += 1
        
        
        self.__outbounds.pop(v)
        self.__inbounds.pop(v)
        self.__vertices.remove(v)
    
    def add_edge(self, v_from:int, v_to:int, e_cost:int):
        edge_id = Edge(v_from, v_to).edge_id

        if self.valid_edge(v_from, v_to) == False:
            raise Exception("Invalid edge!\n")
        
        if self.existing_edge(v_from, v_to) == True:
            raise Exception("Edge already exists!\n")

        self.__outbounds[v_from].append(v_to)
        self.__inbounds[v_to].append(v_from)
        self.__costs[edge_id] = e_cost
        self.__m += 1

    def remove_edge(self, v_from:int, v_to:int):
        if self.existing_edge(v_from, v_to) == False:
            raise Exception("Edge does not exist!\n")

        self.__outbounds[v_from].remove(v_to)
        self.__inbounds[v_to].remove(v_from)
        self.__costs.pop(Edge(v_from, v_to).edge_id)
        self.__m -= 1

    def get_nr_of_vertices(self):
        return len(self.__vertices)
    
    def get_nr_of_edges(self):
        return self.__m
    
    def vertices_iterator(self):
        for vertex in self.__vertices:
            yield vertex

    def edges_iterator(self):
        for edge_id, cost in self.__costs.items():
            v_from, v_to = self.get_endpoints(edge_id)
            yield [v_from, v_to, cost]

    def inbound_iterator(self, vertex_to:int):
        for neighbor in self.__inbounds[vertex_to]:
            yield neighbor

    def outbound_iterator(self, vertex_from:int):
        for neighbor in self.__outbounds[vertex_from]:
            yield neighbor

    def get_in_degree(self, vertex_to:int):
        return len(self.__inbounds[vertex_to])

    def get_out_degree(self, vertex_from:int):
        return len(self.__outbounds[vertex_from])

    def get_cost(self, vertex_from:int, vertex_to:int):
        if self.existing_edge(vertex_from, vertex_to) == False:
            raise Exception("The edge does not exist!\n")
        
        return self.__costs[Edge(vertex_from, vertex_to).edge_id]

    def set_cost(self, vertex_from:int, vertex_to:int, new_cost:int):
        if self.existing_edge(vertex_from, vertex_to) == False:
            raise Exception("The edge does not exist!\n")
        
        self.__costs[Edge(vertex_from, vertex_to).edge_id] = new_cost

    def copy(self):
        return deepcopy(self)

    def existing_vertex(self, v:int):
        return v in self.__vertices

    def valid_vertex(self, v:int):
        return  v >= 0 

    def valid_edge(self, v_from:int, v_to:int):
        return self.valid_vertex(v_from) and self.valid_vertex(v_to) and self.existing_vertex(v_from) and self.existing_vertex(v_to)

    def existing_edge(self, v_from:int, v_to:int):
        return v_from in self.__outbounds.keys() and v_to in self.__outbounds[v_from]

    def get_endpoints(self, edge_id:int):
        v1, v2 = 0, 0
        w = ((8 * edge_id + 1)**0.5 - 1) // 2 

        v2 = edge_id - w * (w + 1) // 2
        v1 = w - v2 

        return (int(v1), int(v2))

    def clear(self):
        self.__n = 0
        self.__m = 0
        self.__vertices.clear()
        self.__inbounds.clear()
        self.__outbounds.clear()

File: first_year/sem2/Graphs/lab/assignment1/C++/Directed_Graph.cpp
#include "Directed_Graph.h"
#include "Edge.h"
#include <iostream>

DirectedGraph::DirectedGraph() : n {0}, m {0}
{}

DirectedGraph::DirectedGraph(int n) : n {n}, m {0}
{
    for (int i = 0; i < n; ++i)
        vertices.push_back(i);
}

DirectedGraph::~DirectedGraph()
{}

DirectedGraph& DirectedGraph::operator = (const DirectedGraph& other_graph) noexcept
{
    in_vertices.clear();
    out_vertices.clear();
    edges.clear();

    n = other_graph.n;
    m = other_graph.m;
    vertices = other_graph.vertices;
    in_vertices = other_graph.in_vertices;
    out_vertices = other_graph.out_vertices;
    edges = other_graph.edges;

    return *this;
}   

int DirectedGraph::GetNrOfVertices() const
{
    return vertices.size();
}

int DirectedGraph::GetNrOfEdges() const
{
    return m;
}

void DirectedGraph::AddEdge(int v1, int v2, int cost)
{
    if (!ValidEdge(v1, v2))
        throw exception("Invalid edge input!\n");

    if (FindEdge(v1, v2) != -1)
        throw exception("This edge already exists!\n");

    in_vertices[v2].push_back(v1);
    out_vertices[v1].push_back(v2);
    edges[Edge(v1, v2).edge_id] = cost;
    ++m;
}

void DirectedGraph::AddVertex(int v)
{
    if (!ValidVertex(v))
        throw exception("Invalid vertex input!\n");

    if (ExistingVertex(v))
        throw exception("Vertex already exists!\n");

    in_vertices[v] = vector<int>();
    out_vertices[v] = vector<int>();
    vertices.push_back(v);
    n = max(n, v + 1);
}

void DirectedGraph::RemoveVertex(int v)
{
    if (!ValidVertex(v) || !ExistingVertex(v))
        throw exception("No such vertex found!\n");

    vertices.erase(find(vertices.begin(), vertices.end(), v));

    // removing all the edges connecting to this vertex
    for (int neighbor : in_vertices[v])
    {
        edges.erase(Edge(neighbor, v).edge_id), --m;
        out_vertices[neighbor].erase(find(out_vertices[neighbor].begin(), out_vertices[neighbor].end(), v));
    }

    for (int neighbor : out_vertices[v])
    {
        edges.erase(Edge(v, neighbor).edge_id), --m;
        in_vertices[neighbor].erase(find(in_vertices[neighbor].begin(), in_vertices[neighbor].end(), v));
    }

    // removing vertex from the graph
    in_vertices.erase(v);
    out_vertices.erase(v);       
}

void DirectedGraph::RemoveEdge(int v1, int v2)
{
    if(!ValidEdge(v1, v2) || !ExistingEdge(v1, v2))
        throw exception("No such edge found!\n");

    edges.erase(Edge(v1, v2).edge_id); // removing the edge

    // removing the neighbor property
    in_vertices[v2].erase(find(in_vertices[v2].begin(), in_vertices[v2].end(), v1));
    out_vertices[v1].erase(find(out_vertices[v1].begin(), out_vertices[v1].end(), v2));
    --m;
}

int DirectedGraph::GetInDegree(int v)
{
    if (ValidVertex(v) && ExistingVertex(v))
        return in_vertices[v].size();

    return -1;
}

int DirectedGraph::GetOutDegree(int v)
{
    if (ValidVertex(v) && ExistingVertex(v))
        return out_vertices[v].size();

    return -1;
}

int DirectedGraph::FindEdge(int v1, int v2)
{
    return edges.count(Edge(v1, v2).edge_id) ? Edge(v1, v2).edge_id : -1;
}

void DirectedGraph::ChangeCost(int v1, int v2, int new_cost)
{
    if (FindEdge(v1, v2) == -1)
        throw exception("Edge not found!\n");

    edges[Edge(v1, v2).edge_id] = new_cost;
}

int DirectedGraph::GetCost(int v1, int v2)
{
    if (FindEdge(v1, v2) == -1)
        throw exception("The edge was not found!\n");

    return edges[Edge(v1, v2).edge_id];
}

// decoding cantor pairings function
pair<int, int> DirectedGraph::GetEndpoints(long long edge_id)
{
    int v1, v2;
    int w = floor((sqrt(8 * edge_id + 1) - 1) / 2);
    v2 = edge_id - (1LL * w * (1LL* w + 1) / 2);
    v1 = w - v2;

    return {v1, v2};
}

void DirectedGraph::ClearGraph()
{
    vertices.clear();
    in_vertices.clear();
    out_vertices.clear();
    edges.clear();
    m = 0;
    n = 0;
}

pair<unordered_map<long long, int>::iterator, unordered_map<long long, int>::iterator> DirectedGraph::IterEdges()
{
    return {edges.begin(), edges.end()};
}

pair<vector<long long>::iterator, vector<long long>::iterator> DirectedGraph::IterOutboundEdges(int v)
{
    outbound_edges.clear();
    
    if (ValidVertex(v) && ExistingVertex(v))
    {
        for (int outbound_vertex : out_vertices[v])
            outbound_edges.push_back(Edge(v, outbound_vertex).edge_id);

        return {outbound_edges.begin(), outbound_edges.end()};
    }

    return {outbound_edges.end(), outbound_edges.end()};
}

pair<vector<long long>::iterator, vector<long long>::iterator> DirectedGraph::IterInboundEdges(int v)
{
    inbound_edges.clear();

    if (ValidVertex(v) && ExistingVertex(v))
    {
        for (int inbound_vertex : in_vertices[v])
            inbound_edges.push_back(Edge(inbound_vertex, v).edge_id);

        return {inbound_edges.begin(), inbound_edges.end()};
    }

    return {inbound_edges.end(), inbound_edges.end()};
}

pair<vector<int>::iterator, vector<int>::iterator> DirectedGraph::IterVertices()
{   
    return {vertices.begin(), vertices.end()};
}

bool DirectedGraph::ValidVertex(int v)
{
    return v >= 0; // true if it's a valid vertex value
}

bool DirectedGraph::ValidEdge(int v1, int v2)
{
    return ValidVertex(v1) && ValidVertex(v2) && ExistingVertex(v1) && ExistingVertex(v2); // true if the vertices are valid and exist in the graph
}

bool DirectedGraph::ExistingVertex(int v)
{
    return find(vertices.begin(), vertices.end(), v) != vertices.end();// true if the vertex does exist in the graph
}

bool DirectedGraph::ExistingEdge(int v1, int v2)
{
    return FindEdge(v1, v2) != -1; // true if the edge exists in the graph already
}